From 6211ae2c90836205615dfd6a6aafe7787f0c214f Mon Sep 17 00:00:00 2001
From: Nan Wu <wnan@google.com>
Date: Mon, 15 Sep 2025 13:59:05 +0000
Subject: [PATCH] RESTRICT AUTOMERGE: When an opaque window has alpha < 0.5,
 don't let it receive touches.

When a window is declared as opaque but the alpha is low, it's likely a fadeout animation or other visual change that should not affect touch handling.

DISABLE_TOPIC_PROTECTOR

Bug: 266433089
Test: inputflinger_tests
Flag: EXEMPT BUGFIX
Change-Id: I4134b2512cf1d0c7bff595db1e011a3357ff130a
---
 services/inputflinger/dispatcher/InputDispatcher.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/services/inputflinger/dispatcher/InputDispatcher.cpp b/services/inputflinger/dispatcher/InputDispatcher.cpp
index 73c5edf2c9..f4f15deacf 100644
--- a/services/inputflinger/dispatcher/InputDispatcher.cpp
+++ b/services/inputflinger/dispatcher/InputDispatcher.cpp
@@ -4918,6 +4918,12 @@ bool InputDispatcher::canWindowReceiveMotionLocked(const sp<WindowInfoHandle>& w
                                                    const MotionEntry& motionEntry) const {
     const WindowInfo& info = *window->getInfo();
 
+    if (info.touchOcclusionMode == TouchOcclusionMode::USE_OPACITY && info.alpha < 0.5f) {
+        LOG(INFO) << "Not sending motion to " << window->getName() << ", window opacity="
+            << info.alpha << " is below the threshold";
+        return false;
+    }
+
     // Skip spy window targets that are not valid for targeted injection.
     if (const auto err = verifyTargetedInjection(window, motionEntry); err) {
         return false;
-- 
2.52.0.rc1.455.g30608eb744-goog

